File: app/integrations/restapiconnections/jira.ts
const fetch = require("node-fetch");

function sleep(ms: number) {
  return new Promise((res) => setTimeout(res, ms));
}

export async function fetchAllJiraIssues(
  projectKey: string, // Have to fetch all relevant proj keys?
  accessToken: string,
): Promise<any[]> {
  let nextPageToken: string | undefined = undefined;
  const maxResults = 5000;
  const allIssues: any[] = [];
  const fields = [
    "summary",
    "description",
    "status",
    "assignee",
    "created",
    "updated",
  ];
  const expand = ["names", "schema", "operations", "changelog"];

  while (true) {
    const queryParams = new URLSearchParams({
      jql: `project = "${projectKey}" ORDER BY created DESC`,
      maxResults: String(maxResults),
      fields: fields.join(","),
      expand: expand.join(","),
    });
    if (nextPageToken) queryParams.append("nextPageToken", nextPageToken);

    let resp: any;
    for (let attempt = 0; attempt < 4; attempt++) {
      resp = await fetch(
        `https://${process.env.JIRA_DOMAIN}/rest/api/3/search/jql?${queryParams.toString()}`,
        {
          method: "GET",
          headers: {
            Authorization: `Bearer ${accessToken}`,
            Accept: "application/json",
          },
        },
      );

      if (resp.status === 429) {
        await sleep(1000);
        continue;
      }

      break;
    }

    if (!resp || !resp.ok) {
      const status = resp ? `${resp.status} ${resp.statusText}` : "no response";
      throw new Error(`Failed to fetch issues: ${status}`);
    }

    const data = await resp.json();
    const issues = data.issues || [];

    // Fetch comments for each issue
    for (const issue of issues) {
      let issueResp: any;
      for (let attempt = 0; attempt < 4; attempt++) {
        issueResp = await fetch(
          `https://${process.env.JIRA_DOMAIN}/rest/api/3/issue/${issue.key}/comment`,
          {
            // Need to add expand query param
            method: "GET",
            headers: {
              Authorization: `Bearer ${accessToken}`,
              Accept: "application/json",
            },
          },
        );

        if (issueResp.status === 429) {
          const retryAfter =
            parseInt(issueResp.headers.get("retry-after") || "1", 10) * 1000;
          await sleep(retryAfter || 1000);
          continue;
        }

        break;
      }

      if (issueResp && issueResp.ok) {
        const commentData = await issueResp.json();
        issue.changelog = commentData.changelog;
        issue.comments = commentData.fields?.comment?.comments || [];
      } else {
        issue.changelog = issue.changelog || null;
        issue.comments = issue.comments || [];
      }
    }

    allIssues.push(...issues);

    if (data.nextPageToken) {
      nextPageToken = data.nextPageToken;
    } else {
      break;
    }
  }

  console.log(allIssues);

  return allIssues;
}

File: backend/integrations/jira.py
import asyncio
from typing import Any, AsyncIterator

import httpx

from integrations.base import BaseIntegration, Document
from settings import settings


class JiraIntegration(BaseIntegration):
    def __init__(self, project_keys: list[str] | None = None, access_token: str | None = None):
        self.project_keys = project_keys or [k.strip() for k in settings.JIRA_PROJECT_KEYS.split(",") if k.strip()]
        self.access_token = access_token or settings.JIRA_ACCESS_TOKEN
        self.domain = settings.JIRA_DOMAIN
        self.headers = {
            "Authorization": f"Bearer {self.access_token}",
            "Accept": "application/json",
        }

    async def _request_with_retry(
        self,
        client: httpx.AsyncClient,
        url: str,
        params: dict[str, str] | None = None,
        max_retries: int = 4,
    ) -> httpx.Response:
        resp: httpx.Response | None = None
        for _attempt in range(max_retries):
            resp = await client.get(url, headers=self.headers, params=params)
            if resp.status_code == 429:
                retry_after = int(resp.headers.get("retry-after", "1"))
                await asyncio.sleep(retry_after)
                continue
            return resp
        if resp is None:
            raise RuntimeError("No response after retries")
        return resp

    async def fetch_documents(self) -> AsyncIterator[Document]:
        fields = ["summary", "description", "status", "assignee", "created", "updated"]
        expand = ["names", "schema", "operations", "changelog"]

        async with httpx.AsyncClient() as client:
            for project_key in self.project_keys:
                next_page_token: str | None = None

                while True:
                    params: dict[str, str] = {
                        "jql": f'project = "{project_key}" ORDER BY created DESC',
                        "maxResults": "100",
                        "fields": ",".join(fields),
                        "expand": ",".join(expand),
                    }
                    if next_page_token:
                        params["nextPageToken"] = next_page_token

                    resp = await self._request_with_retry(
                        client,
                        f"https://{self.domain}/rest/api/3/search/jql",
                        params=params,
                    )
                    resp.raise_for_status()

                    data: dict[str, Any] = resp.json()
                    issues: list[dict[str, Any]] = data.get("issues", [])

                    for issue in issues:
                        # Fetch comments
                        comment_resp = await self._request_with_retry(
                            client,
                            f"https://{self.domain}/rest/api/3/issue/{issue['key']}/comment",
                        )
                        comments: list[dict[str, Any]] = []
                        if comment_resp.status_code == 200:
                            comment_data: dict[str, Any] = comment_resp.json()
                            comments = comment_data.get("comments", [])

                        # Build content
                        description = self._extract_text(issue.get("fields", {}).get("description"))
                        comment_texts = [self._extract_text(c.get("body")) for c in comments]
                        content_parts = [description] + [t for t in comment_texts if t]
                        content = "\n\n".join(content_parts)

                        status = issue.get("fields", {}).get("status", {}).get("name", "")
                        assignee = issue.get("fields", {}).get("assignee", {})
                        assignee_name = assignee.get("displayName", "") if assignee else ""

                        yield Document(
                            source="jira",
                            source_id=issue["key"],
                            title=issue.get("fields", {}).get("summary", ""),
                            content=content,
                            url=f"https://{self.domain}/browse/{issue['key']}",
                            metadata={
                                "project": project_key,
                                "status": status,
                                "assignee": assignee_name,
                                "created": issue.get("fields", {}).get("created", ""),
                                "updated": issue.get("fields", {}).get("updated", ""),
                            },
                        )

                    next_page_token = data.get("nextPageToken")
                    if not next_page_token:
                        break

    def _extract_text(self, adf: Any) -> str:
        """Extract plain text from Atlassian Document Format."""
        if not adf or not isinstance(adf, dict):
            return ""
        parts: list[str] = []
        for node in adf.get("content", []):
            self._walk_adf(node, parts)
        return " ".join(parts)

    def _walk_adf(self, node: Any, parts: list[str]) -> None:
        if not isinstance(node, dict):
            return
        if node.get("type") == "text":
            text = node.get("text", "")
            if text:
                parts.append(text)
        for child in node.get("content", []):
            self._walk_adf(child, parts)
